src/component.js

```diff
@@ -169,338 +126,507 @@ class Component {
+      cntx.addDotenvs(dontenvs.map((v, i, a) => path.join(relativePath, v)));
 
       // third, check component's descriptions
-      if (pair.desc.steps) {
+      if (d.description.steps) {
         // steps' options
         let opts = options.create(this.logger);
-        if (pair.desc.options) {
-          pair.desc.options(null, opts);
+        if (d.description.options) {
+          d.description.options(null, opts);
         }
-        for(const s of pair.desc.steps()) {
+        for (const s of d.description.steps()) {
           // is it our step
           if ((s.id === step) || (step === '*')) {
             // are we meet underyling os, version and other filter's restrictions
-            if (filter.validate(s)) {
+            if (filter.validate(s.filter)) {
               // check if step was already added
-/*               let suffix = [s.id];
-              if (i || (home !== this.getHome())) {
-                suffix.push(`${i}`);
-              }
- */
-              const scriptUid = s.id + '@' + this.getUid([`${i}`]);
-              const scriptName = s.id + '@' + this.getUid([]);
-              if (!r.steps.find( es => es.getUid() === scriptUid )) {
-                r.steps.push( script.create(this.logger, { 
-                  uid: scriptUid,
-                  name: scriptName,
-                  options:opts,
-                  fn: s.script
-                }));
+              /*               let suffix = [s.id];
+                          if (i || (home !== this.getHome())) {
+                            suffix.push(`${i}`);
+                          }
+              */
+              const scriptUuid = s.id + '@' + this.getUuid([`${i}`]);
+              const scriptName = s.id + '@' + this.getUuid([]);
+              if (!r.find(es => es.script.uuid === scriptUuid)) {
+                r.push(
+                  {
+                    script: script.create(this.logger, scriptUuid, scriptName, opts, s.script),
+                    cntx: cntx,
+                  });
               }
             }
           }
         }
       }
     }
     // collect environment variables
-    r.vars = this.getVariables(r.vars);
+    //r.vars = this.getVariables(r.vars);
     return r;
   }
 
-  //
-  async execute(steps, filter, recursive, parallel, params) {
-    this.logger.trace(utils.prefix(this, this.execute.name), utils.quote(this.getId()), 'component executes', steps);
+  /**
+   * Run step based on information from descriptions
+   * params:
+  */
+ async run(steps, filter, recursive, cntx) {
     // collect steps from descs, interits, parents
-    const p = params.clone();
-    p.home = this.getHome();
     for(const step of steps) {
-      const scope2execute = this.findStep(step, filter, p.home, { vars: [], envFiles: [], steps:[] });
+      const list2execute = this.findStep(step, filter, this.home, cntx.clone(), []);
+      const {vars, env} = this.buildEnvironment(this.getVariables());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "vars" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AS1854)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of the unused 'vars' variable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AUnusedVariable)